File: week-03_05/tp/tp_C/tp4/transpose.c
#include "mpi.h"
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  int nb_lines=4;
  int nb_columns=5;
  int nb_lines_t=5;
  int nb_columns_t=4;
  int tag=1000;
  double a[nb_lines][nb_columns];
  double at[nb_lines_t][nb_columns_t];
  int rank,iterl,iterc,size_real;
  MPI_Datatype type_column, type_transpose;
  MPI_Aint size_displacement;
  MPI_Status status;

  /* TODO: Get rank -> rank */

  /* TODO: Type_transpose */

  if (rank == 0) {
    /* Initialisation of A */
    for (iterl=0; iterl<nb_lines; iterl++)
      for (iterc=0; iterc<nb_columns; iterc++)
        a[iterl][iterc] = 1+iterl*nb_columns+iterc;
    printf("Matrix a\n");
    for (iterl=0; iterl<nb_lines;iterl++) {
      for (iterc=0; iterc<nb_columns; iterc++) {
        printf("%4.f ", a[iterl][iterc]);
      }
      printf("\n");
    }

    /* TODO: Send matrix A to process 1 */

  } else {
    /* TODO: Receive in matrix AT with type_transpose */

    printf("Matrice transposee at\n");
    for (iterc=0; iterc<nb_lines_t; iterc++) {
      for (iterl=0; iterl<nb_columns_t;iterl++) {
        printf("%4.f ", at[iterc][iterl]);
      }
      printf("\n");
    }
  }

  /* TODO: Clean MPI types */

  /* TODO: Finalize MPI */

  return 0;
}

File: week-03_05/tp/tp/tp4/transpose.f90
PROGRAM transpose
  USE MPI_F08
  IMPLICIT NONE
  INTEGER, PARAMETER                       :: nb_lines=5,nb_columns=4,&
                                              nb_lines_t=4,nb_columns_t=5,&
                                              tag=1000
  INTEGER                                  :: rank,size_real,i
  TYPE(MPI_Datatype)                       :: type_line,type_transpose
  REAL, DIMENSION(nb_lines,nb_columns)     :: A
  REAL, DIMENSION(nb_lines_t,nb_columns_t) :: AT
  INTEGER(kind=MPI_ADDRESS_KIND)           :: sizeDisplacement
  TYPE(MPI_Status)                         :: status

  ! TODO: Get rank -> rank

  !-- AT matrix  initialisation 
  AT(:,:) = 0.

  ! TODO: datatype transpose

  IF (rank == 0) THEN
    !Initialisation of the A matrix on the 0 process
    A(:,:) = RESHAPE( (/ (i,i=1,nb_lines*nb_columns) /), &
                      (/ nb_lines,nb_columns /) )
    PRINT *,'Matrice A'
    DO i=1,nb_lines
      PRINT *,A(i,:)
    END DO
    ! TODO: Send A matrix to process 1
    
  ELSE
    ! TODO: Reception for the process 1 in AT matrix
    
    PRINT *,'Matrix transpose AT'
    DO i=1,nb_lines_t
      PRINT *,AT(i,:)
    END DO
  END IF

  ! TODO: Clean MPI types

  ! TODO : Finalize MPI

END PROGRAM transpose
